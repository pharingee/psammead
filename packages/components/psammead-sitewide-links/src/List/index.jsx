import React from 'react';
import styled from 'styled-components';
import { string, arrayOf, shape } from 'prop-types';
import { C_SHADOW } from '@bbc/psammead-styles/colours';
import { GEL_SPACING, GEL_SPACING_DBL } from '@bbc/gel-foundations/spacings';
import {
  GEL_GROUP_0_SCREEN_WIDTH_MAX,
  GEL_GROUP_1_SCREEN_WIDTH_MIN,
  GEL_GROUP_2_SCREEN_WIDTH_MAX,
  GEL_GROUP_3_SCREEN_WIDTH_MIN,
  GEL_GROUP_3_SCREEN_WIDTH_MAX,
  GEL_GROUP_4_SCREEN_WIDTH_MIN,
  GEL_GROUP_4_SCREEN_WIDTH_MAX,
  GEL_GROUP_5_SCREEN_WIDTH_MIN,
} from '@bbc/gel-foundations/breakpoints';
import { grid } from '@bbc/psammead-styles/detection';

import Link from '../Link';

// Gets the number of grid rows, taking into account the
// first-child in the grid being separate, on its own row.
const getRowCount = (children, columns) =>
  Math.ceil((React.Children.count(children) - 1) / columns) + 1;

const StyledList = styled.ul`
  border-bottom: 1px solid ${C_SHADOW};
  list-style-type: none;
  margin: 0;
  padding: 0 0 ${GEL_SPACING};
  column-count: 3;

  @supports (${grid}) {
    display: grid;
    grid-auto-flow: column;
  }

  @media (max-width: ${GEL_GROUP_0_SCREEN_WIDTH_MAX}) {
    grid-auto-flow: row;
    column-count: 1;
  }
  @media (min-width: ${GEL_GROUP_1_SCREEN_WIDTH_MIN}) and (max-width: ${GEL_GROUP_2_SCREEN_WIDTH_MAX}) {
    grid-column-gap: ${GEL_SPACING};
    grid-template-columns: repeat(2, 1fr);
    grid-template-rows: repeat(
      ${({ children }) => getRowCount(children, 2)},
      auto
    );
    column-count: 2;
  }
  @media (min-width: ${GEL_GROUP_3_SCREEN_WIDTH_MIN}) and (max-width: ${GEL_GROUP_3_SCREEN_WIDTH_MAX}) {
    grid-column-gap: ${GEL_SPACING_DBL};
    grid-template-columns: repeat(3, 1fr);
    grid-template-rows: repeat(
      ${({ children }) => getRowCount(children, 3)},
      auto
    );
    column-count: 3;
  }
  @media (min-width: ${GEL_GROUP_4_SCREEN_WIDTH_MIN}) and (max-width: ${GEL_GROUP_4_SCREEN_WIDTH_MAX}) {
    grid-column-gap: ${GEL_SPACING_DBL};
    grid-template-columns: repeat(4, 1fr);
    grid-template-rows: repeat(
      ${({ children }) => getRowCount(children, 4)},
      auto
    );
    column-count: 3;
  }
  @media (min-width: ${GEL_GROUP_5_SCREEN_WIDTH_MIN}) {
    grid-column-gap: ${GEL_SPACING_DBL};
    grid-template-columns: repeat(5, 1fr);
    grid-template-rows: repeat(
      ${({ children }) => getRowCount(children, 5)},
      auto
    );
    column-count: 3;
  }
  > li:first-child {
    border-bottom: 1px solid ${C_SHADOW};
    padding: ${GEL_SPACING} 0;
    margin-bottom: ${GEL_SPACING};
    grid-column: 1/-1;
    width: 100%;
    column-span: all;
  }
`;

const StyledListItem = styled.li`
  min-width: 50%;
  column-gap: 1rem;
`;

const List = ({ links }) => (
  <StyledList role="list">
    {links.map(link => (
      <StyledListItem key={link.text} role="listitem">
        <Link text={link.text} href={link.href} />
      </StyledListItem>
    ))}
  </StyledList>
);

const linkPropTypes = shape({
  href: string.isRequired,
  text: string.isRequired,
});

List.propTypes = {
  links: arrayOf(linkPropTypes.isRequired).isRequired,
};

export default List;
