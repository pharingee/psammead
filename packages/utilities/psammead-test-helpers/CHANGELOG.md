# Psammead Test Helpers Changelog

<!-- prettier-ignore -->
| Version | Description |
|---------|-------------|
| 3.0.2 | [PR#2166](https://github.com/bbc/psammead/pull/2166) remove async keyword from test helper to fix regenerator runtime plugin error |
| 3.0.1 | [PR#2123](https://github.com/bbc/psammead/pull/2125) Removes use of async/await |
| 3.0.0 | [PR#2104](https://github.com/bbc/psammead/pull/2104) add function to detect and render helmet for use in snapshots |
| 2.0.2 | [PR#1994](https://github.com/bbc/psammead/pull/1994) Add comment to clarify new shouldMatchSnapshot logic |
| 2.0.1 | [PR#1958](https://github.com/bbc/psammead/pull/1958) Update `shouldMatchSnapshot` to add firstChild logic |
| 2.0.0 | [PR#1917](https://github.com/bbc/psammead/pull/1917) Remove react-test-renderer and shallow rendering |
| 1.3.2 | [PR#1919](https://github.com/bbc/psammead/pull/1919) Accidental version bump |
| 1.3.1 | [PR#1849](https://github.com/bbc/psammead/pull/1849) Update README.md |
| 1.3.0 | [PR#1846](https://github.com/bbc/psammead/pull/1846/) Import test helper functions from simorgh |
| 1.2.1 | [PR#1803](https://github.com/bbc/psammead/pull/1803/) Patches broken links on badges in documentation |
| 1.2.0 | [PR#1794](https://github.com/bbc/psammead/pull/1794) Add david dependency badges |
| 1.1.1 | [PR#1682](https://github.com/bbc/psammead/pull/1682) Move all dev dependencies to top level package.json |
| 1.1.0   | [PR#1233](https://github.com/bbc/psammead/pull/1233) Add ESM modules entry |
| 1.0.2   | [PR#1083](https://github.com/bbc/psammead/pull/1083) Use `jest-styled-components@6.3.3` |
| 1.0.1   | [PR#732](https://github.com/bbc/psammead/pull/732) Use @testing-library/react for generating snapshots |
| 1.0.0   | [PR#679](https://github.com/bbc/psammead/pull/679) Bump version number |
| 0.3.4   | [PR#424](https://github.com/bbc/psammead/pull/424) Add Snyk badge to readme |
| 0.3.3   | [PR#323](https://github.com/BBC/psammead/pull/323) Update readme storybook badge |
| 0.3.2   | [PR#264](https://github.com/BBC/psammead/pull/319) Resolving package-lock issues. |
| 0.3.1   | [PR#264](https://github.com/BBC/psammead/pull/264) Resolving package-lock issues. |
| 0.3.0   | [PR#224](https://github.com/BBC-News/psammead/pull/224) Adds a helper method for testing package export values and uses it to test its own exports. |
| 0.2.1   | [PR#255](https://github.com/BBC-News/psammead/pull/255) Bumps package version to fix what appears to be an erroneously-published version. |
| 0.1.2   | [PR#212](https://github.com/BBC-News/psammead/pull/212) Update package description and README. |
| 0.1.1   | [PR#173](https://github.com/BBC-News/psammead/pull/173) Update PRs welcome link |
| 0.1.0   | [PR#80](https://github.com/BBC-News/psammead/pull/80) Create initial package, pulled in from [simorgh](https://github.com/BBC-News/simorgh). |
