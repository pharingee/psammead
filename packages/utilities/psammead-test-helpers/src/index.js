import { render } from '@testing-library/react';
import 'jest-styled-components';
import deepClone from 'ramda/src/clone';
import renderWithHelmet from './renderWithHelmet';

export const shouldMatchSnapshot = (title, component) => {
  it(title, done => {
    // select the first child to remove the pointless wrapping div from snapshots
    const removeWrappingDiv = container => container.firstChild;
    renderWithHelmet(component).then(({ container }) => {
      const hasOneChild = container.children.length === 1;
      /*
       * if the container has more than one child then it's a component that uses a
       * fragment at the top level so we should not select the first child because it
       * wouldn't snapshot the whole component
       */
      expect(
        hasOneChild ? removeWrappingDiv(container) : container,
      ).toMatchSnapshot();

      done();
    });
  });
};

export const isNull = (title, component) => {
  it(title, () => {
    const { container } = render(component);
    expect(container.firstChild).toBeNull();
  });
};

export const setWindowValue = (key, value) => {
  const windowValue = window[key];
  delete window[key];

  let newValue = value;

  if (value && typeof value === 'object') {
    newValue = {
      ...deepClone(windowValue),
      ...value,
    };
  }

  Object.defineProperty(window, key, {
    value: newValue,
    writable: true,
  });
};

export const resetWindowValue = (key, value) => {
  Object.defineProperty(window, key, {
    value,
    writable: true,
  });
};

export const suppressPropWarnings = warnings => {
  const { expectedWarnings } = window;
  if (expectedWarnings && Array.isArray(expectedWarnings)) {
    window.expectedWarnings = [...expectedWarnings, warnings];
  } else {
    window.expectedWarnings = [warnings];
  }
};

const errorIfMissingKey = (keys, object, message) => {
  keys.forEach(key => {
    if (!(key in object)) {
      throw new Error(`Missing value '${key}' in ${message}.`);
    }
  });
};

const checkKeysExistInBothObjects = (object1, object2, message1, message2) => {
  const object1Keys = Object.keys(object1);
  const object2Keys = Object.keys(object2);

  errorIfMissingKey(object1Keys, object2, message2);
  errorIfMissingKey(object2Keys, object1, message1);
};

const checkTypesOfExports = (
  actualExportsByName,
  actualExports,
  expectedExports,
  utilityName,
) => {
  actualExportsByName.forEach(actualExportName => {
    const actualExportValue = actualExports[utilityName][actualExportName];
    const expectedExport = expectedExports[utilityName][actualExportName];
    const typeCheck = typeof actualExportValue === expectedExport; // eslint-disable-line valid-typeof

    // if this fails it is likely that an export is missing from the unit test expectation
    expect(typeCheck).toBe(true);
  });
};

export const testUtilityPackages = (
  actualExports,
  expectedExports,
  packageName,
) => {
  const actualUtilities = Object.keys(actualExports);

  // check if the actual exported file has defined expected values to match and that the expected values are actually exported
  checkKeysExistInBothObjects(
    actualExports,
    expectedExports,
    `the actual utilities for '${packageName}'`,
    `the expected utilities for '${packageName}'`,
  );

  actualUtilities.forEach(utilityName => {
    const actualExportsByName = Object.keys(actualExports[utilityName]);

    // check if each of the actual exported consts have an expected value to match and that the expected values are actually exported
    checkKeysExistInBothObjects(
      actualExports[utilityName],
      expectedExports[utilityName],
      `the actual export for '${packageName}/${utilityName}'`,
      `the expected export for '${packageName}/${utilityName}'`,
    );

    checkTypesOfExports(
      actualExportsByName,
      actualExports,
      expectedExports,
      utilityName,
    );
  });
};
