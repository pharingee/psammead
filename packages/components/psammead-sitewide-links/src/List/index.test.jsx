import React from 'react';
import { shouldMatchSnapshot } from '@bbc/psammead-test-helpers';
import List from './index';

describe(`List`, () => {
  const link = {
    href: 'https://www.bbc.co.uk/news',
    text: 'Link',
  };

  const getLinks = count =>
    new Array(count).fill(link).map((linkItem, index) => ({
      ...linkItem,
      text: `${linkItem.text}${index}`,
    }));

  shouldMatchSnapshot(
    'should render correctly with 7 items',
    <List links={getLinks(7)} />,
  );

  shouldMatchSnapshot(
    'should render correctly with 8 items',
    <List links={getLinks(8)} />,
  );
});
