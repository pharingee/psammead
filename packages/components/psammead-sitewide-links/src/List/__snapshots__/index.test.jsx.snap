// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`List should render correctly with 7 items 1`] = `
.c3 {
  padding: 0.5rem 0 0.5rem;
  color: #FFFFFF;
  font-weight: 700;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
}

.c2:hover .c4,
.c2:focus .c4 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 {
  border-bottom: 1px solid #3F3F42;
  list-style-type: none;
  margin: 0;
  padding: 0 0 0.5rem;
}

.c0 > li:first-child {
  border-bottom: 1px solid #3F3F42;
  padding: 0.5rem 0;
  margin-bottom: 0.5rem;
  grid-column: 1/-1;
}

.c1 {
  min-width: 50%;
}

@supports (grid-template-columns:fit-content(200px)) {
  .c0 {
    display: grid;
    grid-auto-flow: column;
  }
}

@media (max-width:14.9375rem) {
  .c0 {
    grid-auto-flow: row;
  }
}

@media (min-width:15rem) and (max-width:37.4375rem) {
  .c0 {
    grid-column-gap: 0.5rem;
    grid-template-columns: repeat(2,1fr);
    grid-template-rows: repeat( 4,auto );
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c0 {
    grid-column-gap: 1rem;
    grid-template-columns: repeat(3,1fr);
    grid-template-rows: repeat( 3,auto );
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column-gap: 1rem;
    grid-template-columns: repeat(4,1fr);
    grid-template-rows: repeat( 3,auto );
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column-gap: 1rem;
    grid-template-columns: repeat(5,1fr);
    grid-template-rows: repeat( 3,auto );
  }
}

@supports not (grid-template-columns:fit-content(200px)) {
  .c0 > li:first-child {
    width: 100%;
  }
}

@supports not (grid-template-columns:fit-content(200px)) {
  .c1 {
    display: inline-block;
  }
}

<ul
  class="c0"
  role="list"
>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link0
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link1
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link2
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link3
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link4
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link5
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link6
      </span>
    </a>
  </li>
</ul>
`;

exports[`List should render correctly with 8 items 1`] = `
.c3 {
  padding: 0.5rem 0 0.5rem;
  color: #FFFFFF;
  font-weight: 700;
  -webkit-text-decoration: none;
  text-decoration: none;
  display: block;
}

.c2:hover .c4,
.c2:focus .c4 {
  -webkit-text-decoration: underline;
  text-decoration: underline;
}

.c0 {
  border-bottom: 1px solid #3F3F42;
  list-style-type: none;
  margin: 0;
  padding: 0 0 0.5rem;
}

.c0 > li:first-child {
  border-bottom: 1px solid #3F3F42;
  padding: 0.5rem 0;
  margin-bottom: 0.5rem;
  grid-column: 1/-1;
}

.c1 {
  min-width: 50%;
}

@supports (grid-template-columns:fit-content(200px)) {
  .c0 {
    display: grid;
    grid-auto-flow: column;
  }
}

@media (max-width:14.9375rem) {
  .c0 {
    grid-auto-flow: row;
  }
}

@media (min-width:15rem) and (max-width:37.4375rem) {
  .c0 {
    grid-column-gap: 0.5rem;
    grid-template-columns: repeat(2,1fr);
    grid-template-rows: repeat( 5,auto );
  }
}

@media (min-width:37.5rem) and (max-width:62.9375rem) {
  .c0 {
    grid-column-gap: 1rem;
    grid-template-columns: repeat(3,1fr);
    grid-template-rows: repeat( 4,auto );
  }
}

@media (min-width:63rem) and (max-width:80rem) {
  .c0 {
    grid-column-gap: 1rem;
    grid-template-columns: repeat(4,1fr);
    grid-template-rows: repeat( 3,auto );
  }
}

@media (min-width:80rem) {
  .c0 {
    grid-column-gap: 1rem;
    grid-template-columns: repeat(5,1fr);
    grid-template-rows: repeat( 3,auto );
  }
}

@supports not (grid-template-columns:fit-content(200px)) {
  .c0 > li:first-child {
    width: 100%;
  }
}

@supports not (grid-template-columns:fit-content(200px)) {
  .c1 {
    display: inline-block;
  }
}

<ul
  class="c0"
  role="list"
>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link0
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link1
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link2
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link3
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link4
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link5
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link6
      </span>
    </a>
  </li>
  <li
    class="c1"
    role="listitem"
  >
    <a
      class="c2 c3"
      href="https://www.bbc.co.uk/news"
    >
      <span
        class="c4"
      >
        Link7
      </span>
    </a>
  </li>
</ul>
`;
